src/lib/proof-system/zkprogram.ts

```diff
@@ -803,44 +845,53 @@ function picklesRuleFromFunction(
   ): ReturnType<Pickles.Rule['main']> {
     let { witnesses: argsWithoutPublicInput, inProver } = snarkContext.get();
     assert(!(inProver && argsWithoutPublicInput === undefined));
+
+    // witness private inputs and declare input proofs
+    let id = ZkProgramContext.enter();
     let finalArgs = [];
-    let proofs: {
-      Proof: Subclass<typeof ProofBase<any, any>>;
-      proof: ProofBase<any, any>;
-    }[] = [];
-    let previousStatements: Pickles.Statement<FieldVar>[] = [];
     for (let i = 0; i < args.length; i++) {
-      let type = args[i];
       try {
+        let type = args[i];
         let value = Provable.witness(type, () => {
           return argsWithoutPublicInput?.[i] ?? ProvableType.synthesize(type);
         });
         finalArgs[i] = value;
 
-        for (let proof of extractProofs(value)) {
-          let Proof = proof.constructor as Subclass<typeof ProofBase<any, any>>;
-          proofs.push({ Proof, proof });
-          let fields = proof.publicFields();
-          let input = MlFieldArray.to(fields.input);
-          let output = MlFieldArray.to(fields.output);
-          previousStatements.push(MlPair(input, output));
-        }
+        extractProofs(value).forEach((proof) => proof.declare());
       } catch (e: any) {
+        ZkProgramContext.leave(id);
         e.message = `Error when witnessing in ${methodName}, argument ${i}: ${e.message}`;
         throw e;
       }
     }
-    let result: {
-      publicOutput?: any;
-      auxiliaryOutput?: any;
-    };
-    if (publicInputType === Undefined || publicInputType === Void) {
-      result = (await func(...finalArgs)) as any;
-    } else {
-      let input = fromFieldVars(publicInputType, publicInput);
-      result = (await func(input, ...finalArgs)) as any;
+
+    // run the user circuit
+    let result: { publicOutput?: any; auxiliaryOutput?: any };
+    let proofs: DeclaredProof[];
+
+    try {
+      if (publicInputType === Undefined || publicInputType === Void) {
+        result = (await func(...finalArgs)) as any;
+      } else {
+        let input = fromFieldVars(publicInputType, publicInput);
+        result = (await func(input, ...finalArgs)) as any;
+      }
+      proofs = ZkProgramContext.getDeclaredProofs();
```

Comment:
instead of collecting all proofs inside this function as before, we get them from the new global context `ZkProgramContext`, which is also written to by `Proof.declare()`